File: android/app/src/main/java/my/ssdid/wallet/feature/notifications/NotificationsScreen.kt
package my.ssdid.wallet.feature.notifications

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import my.ssdid.wallet.domain.notify.LocalNotification
import my.ssdid.wallet.domain.notify.LocalNotificationStorage
import my.ssdid.wallet.ui.theme.*
import javax.inject.Inject

@HiltViewModel
class NotificationsViewModel @Inject constructor(
    private val storage: LocalNotificationStorage
) : ViewModel() {

    val notifications = storage.allNotifications
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    fun markAsRead(id: String) {
        viewModelScope.launch { storage.markAsRead(id) }
    }

    fun markAllAsRead() {
        viewModelScope.launch { storage.markAllAsRead() }
    }

    fun delete(id: String) {
        viewModelScope.launch { storage.delete(id) }
    }
}

@Composable
fun NotificationsScreen(
    onBack: () -> Unit,
    viewModel: NotificationsViewModel = hiltViewModel()
) {
    val notifications by viewModel.notifications.collectAsState()
    val sorted = notifications.sortedByDescending { it.receivedAt }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(BgPrimary)
            .statusBarsPadding()
    ) {
        // Header
        Row(
            Modifier
                .fillMaxWidth()
                .padding(start = 8.dp, end = 20.dp, top = 12.dp, bottom = 12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            IconButton(onClick = onBack) {
                Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back", tint = TextPrimary)
            }
            Spacer(Modifier.width(4.dp))
            Text("Notifications", style = MaterialTheme.typography.titleLarge, modifier = Modifier.weight(1f))
            if (notifications.any { !it.isRead }) {
                TextButton(onClick = { viewModel.markAllAsRead() }) {
                    Text("Mark All Read", color = Accent, fontSize = 13.sp)
                }
            }
        }

        if (sorted.isEmpty()) {
            // Empty state
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(32.dp),
                contentAlignment = Alignment.Center
            ) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Box(
                        modifier = Modifier
                            .size(72.dp)
                            .clip(RoundedCornerShape(20.dp))
                            .background(BgCard),
                        contentAlignment = Alignment.Center
                    ) {
                        Icon(
                            Icons.Default.Notifications,
                            contentDescription = null,
                            modifier = Modifier.size(32.dp),
                            tint = TextSecondary
                        )
                    }
                    Spacer(Modifier.height(16.dp))
                    Text("No notifications", style = MaterialTheme.typography.headlineSmall, color = TextPrimary)
                    Spacer(Modifier.height(4.dp))
                    Text(
                        "Notifications from services will appear here",
                        color = TextSecondary,
                        fontSize = 14.sp
                    )
                }
            }
        } else {
            LazyColumn(
                Modifier.padding(horizontal = 20.dp),
                verticalArrangement = Arrangement.spacedBy(6.dp)
            ) {
                items(sorted, key = { it.id }) { notification ->
                    NotificationRow(
                        notification = notification,
                        onTap = { viewModel.markAsRead(notification.id) },
                        onDelete = { viewModel.delete(notification.id) }
                    )
                }
                item { Spacer(Modifier.height(20.dp)) }
            }
        }
    }
}

@Composable
private fun NotificationRow(
    notification: LocalNotification,
    onTap: () -> Unit,
    onDelete: () -> Unit
) {
    val dismissState = rememberSwipeToDismissBoxState(
        confirmValueChange = { value ->
            if (value == SwipeToDismissBoxValue.EndToStart) {
                onDelete()
                true
            } else false
        }
    )

    SwipeToDismissBox(
        state = dismissState,
        backgroundContent = {
            Box(
                Modifier
                    .fillMaxSize()
                    .clip(RoundedCornerShape(12.dp))
                    .background(Danger)
                    .padding(horizontal = 20.dp),
                contentAlignment = Alignment.CenterEnd
            ) {
                Text("Delete", color = BgPrimary, fontWeight = FontWeight.SemiBold)
            }
        },
        enableDismissFromStartToEnd = false
    ) {
        Card(
            Modifier
                .fillMaxWidth()
                .clickable(onClick = onTap),
            shape = RoundedCornerShape(12.dp),
            colors = CardDefaults.cardColors(containerColor = BgCard)
        ) {
            Row(
                Modifier.padding(14.dp),
                verticalAlignment = Alignment.Top
            ) {
                // Unread dot
                Box(
                    modifier = Modifier
                        .padding(top = 6.dp)
                        .size(8.dp)
                        .clip(CircleShape)
                        .background(if (!notification.isRead) Accent else BgCard)
                )
                Spacer(Modifier.width(12.dp))
                Column(Modifier.weight(1f)) {
                    if (notification.identityName != null) {
                        Text(
                            notification.identityName,
                            fontSize = 12.sp,
                            color = TextTertiary,
                            fontWeight = FontWeight.Medium
                        )
                        Spacer(Modifier.height(2.dp))
                    }
                    Text(
                        notification.payload,
                        fontSize = 14.sp,
                        color = TextPrimary,
                        maxLines = 3
                    )
                }
                Spacer(Modifier.width(8.dp))
                Text(
                    relativeTime(notification.receivedAt),
                    fontSize = 12.sp,
                    color = TextTertiary
                )
            }
        }
    }
}

private fun relativeTime(isoTimestamp: String): String {
    if (isoTimestamp.isBlank()) return ""
    return try {
        val instant = java.time.Instant.parse(isoTimestamp)
        val now = java.time.Instant.now()
        val minutes = java.time.Duration.between(instant, now).toMinutes()
        when {
            minutes < 1 -> "now"
            minutes < 60 -> "${minutes}m ago"
            minutes < 1440 -> "${minutes / 60}h ago"
            minutes < 2880 -> "Yesterday"
            else -> "${minutes / 1440}d ago"
        }
    } catch (_: Exception) {
        ""
    }
}

File: ios/SsdidWallet/Feature/Notifications/NotificationsScreen.swift
import SwiftUI

struct NotificationsScreen: View {
    @Environment(AppRouter.self) private var router
    @EnvironmentObject private var services: ServiceContainer

    var body: some View {
        let storage = services.localNotificationStorage
        let isoFormatter = ISO8601DateFormatter()
        let sorted = storage.notifications.sorted {
            let d0 = isoFormatter.date(from: $0.receivedAt) ?? .distantPast
            let d1 = isoFormatter.date(from: $1.receivedAt) ?? .distantPast
            return d0 > d1
        }

        VStack(spacing: 0) {
            // Header
            HStack {
                Button { router.pop() } label: {
                    Image(systemName: "chevron.left")
                        .font(.system(size: 18, weight: .semibold))
                        .foregroundStyle(Color.textPrimary)
                }
                Spacer().frame(width: 12)
                Text("Notifications")
                    .font(.ssdidTitle)
                    .foregroundStyle(Color.textPrimary)
                Spacer()
                if storage.unreadCount > 0 {
                    Button {
                        storage.markAllAsRead()
                    } label: {
                        Text("Mark All Read")
                            .font(.system(size: 13))
                            .foregroundStyle(Color.ssdidAccent)
                    }
                }
            }
            .padding(.horizontal, 20)
            .padding(.vertical, 12)

            if sorted.isEmpty {
                Spacer()
                VStack(spacing: 0) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 20)
                            .fill(Color.bgCard)
                            .frame(width: 72, height: 72)
                        Image(systemName: "bell")
                            .font(.system(size: 32))
                            .foregroundStyle(Color.textSecondary)
                    }
                    Spacer().frame(height: 16)
                    Text("No notifications")
                        .font(.ssdidHeadline)
                        .foregroundStyle(Color.textPrimary)
                    Spacer().frame(height: 4)
                    Text("Notifications from services will appear here")
                        .font(.system(size: 14))
                        .foregroundStyle(Color.textSecondary)
                        .multilineTextAlignment(.center)
                }
                Spacer()
            } else {
                ScrollView {
                    LazyVStack(spacing: 6) {
                        ForEach(sorted) { notification in
                            notificationRow(notification, storage: storage)
                        }
                    }
                    .padding(.horizontal, 20)
                    .padding(.bottom, 20)
                }
            }
        }
        .background(Color.bgPrimary)
    }

    @ViewBuilder
    private func notificationRow(_ notification: LocalNotification, storage: LocalNotificationStorage) -> some View {
        HStack(alignment: .top, spacing: 12) {
            // Unread dot
            Circle()
                .fill(notification.isRead ? Color.clear : Color.ssdidAccent)
                .frame(width: 8, height: 8)
                .padding(.top, 6)

            VStack(alignment: .leading, spacing: 2) {
                if let name = notification.identityName {
                    Text(name)
                        .font(.system(size: 12, weight: .medium))
                        .foregroundStyle(Color.textTertiary)
                }
                Text(notification.payload)
                    .font(.system(size: 14))
                    .foregroundStyle(Color.textPrimary)
                    .lineLimit(3)
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            Text(relativeTime(notification.receivedAt))
                .font(.system(size: 12))
                .foregroundStyle(Color.textTertiary)
        }
        .padding(14)
        .background(Color.bgCard)
        .cornerRadius(12)
        .onTapGesture {
            storage.markAsRead(notification.id)
        }
    }

    private func relativeTime(_ isoTimestamp: String) -> String {
        guard !isoTimestamp.isEmpty else { return "" }
        let formatter = ISO8601DateFormatter()
        guard let date = formatter.date(from: isoTimestamp) else { return "" }
        let minutes = Int(Date().timeIntervalSince(date) / 60)
        switch minutes {
        case ..<1: return "now"
        case ..<60: return "\(minutes)m ago"
        case ..<1440: return "\(minutes / 60)h ago"
        case ..<2880: return "Yesterday"
        default: return "\(minutes / 1440)d ago"
        }
    }
}
